master/internal/hpimportance/manager.go

```diff
@@ -0,0 +1,284 @@
+		metricData.Error = msg.err
+		setMetricHPImportance(&hpi, metricData, msg.metricName, msg.metricType)
+		err = m.db.SetHPImportance(msg.experimentID, hpi)
+		if err != nil {
+			ctx.Log().Errorf("error writing hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+	case workCompleted:
+		m.state[msg.experimentID] = stateRecord{
+			lastResult:   time.Now(),
+			lastProgress: m.state[msg.experimentID].lastProgress,
+		}
+		hpi, err := m.db.GetHPImportance(msg.experimentID)
+		if err != nil {
+			ctx.Log().Errorf("error retrieving hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+		metricData := getMetricHPImportance(hpi, msg.metricName, msg.metricType)
+		metricData.ExperimentProgress = msg.progress
+		metricData.HpImportance = msg.results
+		switch metricData.Status {
+		case Pending:
+			// Do nothing - this means another startWork message was already sent
+		case InProgress:
+			metricData.Status = Complete
+		default:
+			ctx.Log().Warnf("work was completed for a metric with an unexpected state")
+		}
+		setMetricHPImportance(&hpi, metricData, msg.metricName, msg.metricType)
+		err = m.db.SetHPImportance(msg.experimentID, hpi)
+		if err != nil {
+			ctx.Log().Errorf("error writing hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+	default:
+		ctx.Log().Errorf("unknown message received by hyperparameter importance manager: %v!",
+			ctx.Message())
+	}
+	return nil
+}
+
+func (m *manager) getChild(ctx *actor.Context, experimentID int) *actor.Ref {
+	var result *actor.Ref
+	/*
+		Currently each experiment gets their own actor, that will persist even after the work
+		for an experiment may be complete. But the child actor's don't need to maintain any state
+		between tasks, so if this becomes a problem in practice we could use a threadpool model
+		where the manager maintains a queue of tasks, workers .Ask() the manager for a task (and it
+		may return a task to sleep for a time), and the manager decides when to spawn or scale down
+		workers. Or each authenticated user can have their own actor, for improved multi-tenancy.
+	*/
+	result = ctx.Child(experimentID)
+	if result == nil {
+		w := newWorker(m.db, ctx.Self())
+		result, _ = ctx.ActorOf(experimentID, w)
+	}
+	return result
+}
+
+func (m *manager) triggerDefaultWork(ctx *actor.Context, experimentID int) {
+	child := m.getChild(ctx, experimentID)
+
+	hpi, err := m.db.GetHPImportance(experimentID)
+	if err != nil {
+		ctx.Log().Errorf("error retrieving hyperparameter importance state: %s", err.Error())
+		return
+	}
+
+	config, err := m.db.ExperimentConfig(experimentID)
+	if err != nil {
+		ctx.Log().Errorf("error retrieving experiment config: %s", err.Error())
+		return
+	}
+
+	loss := "loss"
+	triggerForLoss := false
+	lossHpi := getMetricHPImportance(hpi, loss, Training)
+	if lossHpi.Status != Pending {
+		triggerForLoss = true
+	}
+	lossHpi.Status = Pending
+	setMetricHPImportance(&hpi, lossHpi, loss, Training)
```

Comment:
What's abstracted into the functions is all the switching between using hpi.TrainingMetrics or hpi.ValidationMetrics. But yes, get has no need to modify the dictionary (especially since the manager is single-threaded) as the set doesn't need it to be initialized to an empty value or anything. So that's out. Also testing the move to have these be functions on hpi and I'll upload that shortly. I actually really like that...